File: v8-core/src/audit/authority.rs
//! Authority Audit & Receipt Chain Validator (D-132, Rule 28).

use crate::authority::{Authority, ClaimValue, ConstitutionalViolation, DecisionAuthority};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct AuthorityAuditReport {
    pub passed: bool,
    pub receipts_checked: usize,
    pub violations: Vec<String>,
}

pub struct AuthorityAuditor;

impl AuthorityAuditor {
    /// Validates that an output authority does not exceed the minimum of parent authorities.
    pub fn audit_monotonicity(
        output_auth: &Authority,
        input_auths: &[Authority],
    ) -> Result<(), ConstitutionalViolation> {
        if input_auths.is_empty() {
            return Ok(());
        }

        let mut min_auth = input_auths[0];
        for auth in &input_auths[1..] {
            min_auth = min_auth.meet(auth);
        }

        if (output_auth.evidence as u8) > (min_auth.evidence as u8) {
            return Err(ConstitutionalViolation::AuthorityEscalationAttempted {
                attempted_evidence: output_auth.evidence,
                source_evidence: min_auth.evidence,
            });
        }
        if (output_auth.decision as u8) > (min_auth.decision as u8) {
            return Err(ConstitutionalViolation::ClaimBlocked {
                required_authority: format!("{:?}", min_auth.decision),
                provided_authority: format!("{:?}", output_auth.decision),
                reason: "DECISION_AUTHORITY_ESCALATION".to_string(),
            });
        }
        if (output_auth.realization as u8) > (min_auth.realization as u8) {
            return Err(ConstitutionalViolation::ClaimBlocked {
                required_authority: format!("{:?}", min_auth.realization),
                provided_authority: format!("{:?}", output_auth.realization),
                reason: "REALIZATION_STATUS_ESCALATION".to_string(),
            });
        }

        Ok(())
    }

    /// Audits a collection of claim values to verify receipt presence and authority contracts.
    pub fn audit_claims<T>(claims: &[ClaimValue<T>]) -> AuthorityAuditReport {
        let mut violations = Vec::new();
        for (i, c) in claims.iter().enumerate() {
            if c.receipt_id.is_empty() {
                violations.push(format!("Claim at index {i} has empty receipt_id"));
            }
            if c.authority.decision >= DecisionAuthority::PortfolioAuthorized && c.receipt_id.len() < 16 {
                violations.push(format!("Claim at index {i} has invalid cryptographic receipt"));
            }
        }

        AuthorityAuditReport {
            passed: violations.is_empty(),
            receipts_checked: claims.len(),
            violations,
        }
    }
}

File: v8-next/src/v8_next/audit/authority.py
"""Authority audit and receipt-chain validation — port of v8-core/src/audit/authority.rs.

The Rust auditor compares an output ``Authority {evidence, decision, realization}`` against
the minimum of its inputs. In this codebase the authority ladder is the statutory claim
class (``evaluation.claims.StatutoryClaimClass``), so:

* rungs up to ``SIMULATED_CASHFLOW`` are the **evidence** segment,
* ``REALIZED_CASHFLOW`` is the **realization** segment,
* rungs above it (``READY_NOT_CLAIMED``, ``SUPPORTED_EDGE``) are the **decision** segment.

An output may never sit higher on that ladder than the lowest input that produced it.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from enum import StrEnum

from v8_next.evaluation.claims import StatutoryClaimClass, StatutoryClaimRecord

#: the ladder, weakest first. Index = authority level.
AUTHORITY_LADDER: tuple[StatutoryClaimClass, ...] = (
    StatutoryClaimClass.DiagnosticSignal,
    StatutoryClaimClass.CounterfactualPotential,
    StatutoryClaimClass.RecoverableRegret,
    StatutoryClaimClass.SimulatedCashflow,
    StatutoryClaimClass.RealizedCashflow,
    StatutoryClaimClass.ReadyNotClaimed,
    StatutoryClaimClass.SupportedEdge,
)

_EVIDENCE_TOP = AUTHORITY_LADDER.index(StatutoryClaimClass.SimulatedCashflow)
_REALIZATION = AUTHORITY_LADDER.index(StatutoryClaimClass.RealizedCashflow)

#: a claim at or above this rung needs a real cryptographic receipt, not a placeholder
RECEIPT_REQUIRED_FROM = AUTHORITY_LADDER.index(StatutoryClaimClass.RealizedCashflow)
MIN_RECEIPT_LENGTH = 16


class AuthoritySegment(StrEnum):
    EVIDENCE = "EVIDENCE"
    REALIZATION = "REALIZATION"
    DECISION = "DECISION"


def authority_level(claim_class: StatutoryClaimClass) -> int:
    return AUTHORITY_LADDER.index(claim_class)


def authority_segment(claim_class: StatutoryClaimClass) -> AuthoritySegment:
    level = authority_level(claim_class)
    if level <= _EVIDENCE_TOP:
        return AuthoritySegment.EVIDENCE
    if level <= _REALIZATION:
        return AuthoritySegment.REALIZATION
    return AuthoritySegment.DECISION


class ConstitutionalViolation(Exception):
    """Raised when an output claims more authority than its inputs license (Rule 28)."""

    def __init__(self, code: str, message: str) -> None:
        super().__init__(f"{code}: {message}")
        self.code = code
        self.message = message


@dataclass(frozen=True)
class AuthorityAuditReport:
    passed: bool
    receipts_checked: int
    violations: list[str] = field(default_factory=list)


class AuthorityAuditor:
    """Port of ``v8_core::audit::AuthorityAuditor``."""

    @staticmethod
    def audit_monotonicity(
        output_claim: StatutoryClaimRecord,
        input_claims: list[StatutoryClaimRecord],
    ) -> None:
        """An output may not exceed the minimum authority of the inputs that produced it."""
        if not input_claims:
            return
        min_claim = min(input_claims, key=lambda c: authority_level(c.claim_class))
        output_level = authority_level(output_claim.claim_class)
        min_level = authority_level(min_claim.claim_class)
        if output_level <= min_level:
            return
        segment = authority_segment(output_claim.claim_class)
        code = {
            AuthoritySegment.EVIDENCE: "AUTHORITY_ESCALATION_ATTEMPTED",
            AuthoritySegment.REALIZATION: "REALIZATION_STATUS_ESCALATION",
            AuthoritySegment.DECISION: "DECISION_AUTHORITY_ESCALATION",
        }[segment]
        raise ConstitutionalViolation(
            code,
            f"output {output_claim.claim_class.value} (level {output_level}) exceeds the "
            f"lowest input {min_claim.claim_class.value} (level {min_level})",
        )

    @staticmethod
    def audit_claims(claims: list[StatutoryClaimRecord]) -> AuthorityAuditReport:
        """Every claim must carry its provenance; a strong claim needs a real receipt."""
        violations: list[str] = []
        for index, claim in enumerate(claims):
            if not claim.parent_receipt_hashes:
                violations.append(f"Claim at index {index} has no parent receipt")
            if (
                authority_level(claim.claim_class) >= RECEIPT_REQUIRED_FROM
                and len(claim.signature) < MIN_RECEIPT_LENGTH
            ):
                violations.append(f"Claim at index {index} has invalid cryptographic receipt")
        return AuthorityAuditReport(
            passed=not violations, receipts_checked=len(claims), violations=violations
        )
